### PR TITLE
add roster sync opt out when user syncs from an LMS and upgrade sync dialog to TypeScript

### PR DESCRIPTION
**Pull request by stephenliang**
## Links

- jira: [Add opt-out button in sync course pathway](https://codedotorg.atlassian.net/browse/P20-704)

## Testing story

1. Added unit tests and backported missing test cases for the sync dialog
2. Added ruby unit tests for controller
3. Ran manual sync from canvas, see below video

[LMS Opt Out Button.webm](https://github.com/code-dot-org/code-dot-org/assets/538214/6bffd882-8464-4ce6-89a5-69b1d48f5b68)

## PR Checklist:

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [x] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/test/controllers/api/v1/users_controller_test.rb [hunk 1]
```diff
@@ -87,6 +87,35 @@ class Api::V1::UsersControllerTest < ActionController::TestCase
     refute @user.show_progress_table_v2
   end
 
+  test 'a post request to disable_lti_roster_sync updates lti_roster_sync_enabled' do
+    teacher = create :teacher
+    teacher.lti_roster_sync_enabled = true
```
Comment (carl-codeorg) [inline dashboard/test/controllers/api/v1/users_controller_test.rb]: Tiny nitpick - I think you can just do `teacher = create :teacher, lti_roster_sync_enabled: true` to make it a one-liner.

File: apps/src/lib/ui/lti/sync/types.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export type LtiSectionMap = {[sectionId: string]: LtiSection};
+
+export interface LtiSectionSyncResult {
+  all?: LtiSectionMap;
+  updated?: LtiSectionMap;
+  error?: string;
+}
+
+export interface LtiSectionSyncDialogProps {
```
Comment (carl-codeorg) [inline apps/src/lib/ui/lti/sync/types.ts:14]: With the props defined here, do we still need `LtiSectionSyncDialog.propTypes` in `LtiSectionSyncDialog.tsx`?

File: apps/src/lib/ui/lti/sync/LtiSectionSyncDialog.tsx [hunk 3]
```diff
@@ -1,24 +1,30 @@
...
-  const SYNC_RESULT_VIEW = 'syncResult';
-  const SPINNER_VIEW = 'spinner';
-  const ERROR_VIEW = 'error';
-  const {syncResult, onClose} = props;
-  const initialView = syncResult.error ? ERROR_VIEW : SYNC_RESULT_VIEW;
-  const [currentView, setCurrentView] = useState(initialView);
+export default function LtiSectionSyncDialog({
+  syncResult,
+  onClose,
+  disableRosterSyncEnabled,
```
Comment (carl-codeorg) [inline apps/src/lib/ui/lti/sync/LtiSectionSyncDialog.tsx]: The name `disableRosterSyncEnabled` is a bit confusing to me. Can you explain what this prop does?

File: apps/src/lib/ui/lti/sync/types.ts (hunk 2 above)
Comment (stephenliang) [reply apps/src/lib/ui/lti/sync/types.ts:14]: Unfortunately the PropTypes are used by other JSX components. Unless we also migrate those to Typescript, we'd have to continue to have the PropTypes. I didn't migrate those to TS to keep the PR size fairly small and to keep the change fairly lightweight. We can circle back and once the other components are converted to TSX, these PropTypes will be removed as there wouldn't be any dependent usages anymore.

File: apps/src/lib/ui/lti/sync/LtiSectionSyncDialog.tsx (hunk 3 above)
Comment (stephenliang) [reply apps/src/lib/ui/lti/sync/LtiSectionSyncDialog.tsx]: `disableRosterSyncEnabled` enables the button to turn off roster sync. Per a prior discussion, we do not want to show the "Disable roster sync" button when clicking "Sync Courses" but we do want to show "Disable roster sync" when syncing courses via SSO sign in.

I renamed the prop to `disableRosterSyncButtonEnabled` be more clear

**Review by carl-codeorg — APPROVED** [review]
This LGTM! :shipit:

**Pull request merged**